internal/memory/pool.go

```diff
@@ -0,0 +1,323 @@
+		pa.pools[clsType] = cp
+	}
+	// create 4 smaller regions
+	for i := uint64(0); i < 4; i++ {
+		offset := nextReg.offset + i*clsSize
+		reg := &region{
+			parent: nextReg,
+			class:  clsType,
+			offset: offset,
+		}
+		cp.free[offset] = reg
+	}
+	return nil
+}
+
+// Release marks a memory region of class `memCls` and offset `offset` as free and tries to merge smaller regions into
+// a bigger one
+func (pa *PoolAllocator) Release(memCls classType, offset uint64) error {
+	mp := pa.pools[memCls]
+	if mp == nil {
+		return ErrEmptyPoolOperation
+	}
+
+	n, err := pa.markFree(memCls, offset)
+	if err != nil {
+		return err
+	}
+
+	if err := pa.merge(n.parent); err != nil {
+		if err != ErrEarlyMerge {
+			return err
+		}
+	}
+	return nil
+}
+
+func (pa *PoolAllocator) merge(parent *region) error {
+	// nothing to merge
+	if parent == nil {
+		return nil
+	}
+
+	parentCls := parent.class
+	// nothing to do, cannot merge the smallest class
+	if parentCls == 0 {
+		return nil
+	}
+	parentPool := pa.pools[parentCls]
+	// cannot merge an empty pool
+	if parentPool == nil {
+		return ErrEmptyPoolOperation
+	}
+
+	childCls := parentCls - 1
+	childPool := pa.pools[childCls]
+	// no child nodes to merge, try merge parent
+	if childPool == nil {
+		return pa.merge(parent.parent)
+	}
+
+	childSize, err := GetMemoryClassSize(childCls)
+	if err != nil {
+		return err
+	}
+
+	// check if all of the child nodes are free
+	var children []*region
+	for i := uint64(0); i < 4; i++ {
+		child, free := childPool.free[parent.offset+i*childSize]
+		if !free {
+			return ErrEarlyMerge
+		}
+		children = append(children, child)
+	}
+
+	// at this point all of the child nodes will be free and we can merge
+	for _, child := range children {
+		delete(childPool.free, child.offset)
+	}
+
+	if _, err := pa.markFree(parent.class, parent.offset); err != nil {
+		return err
+	}
+
+	return pa.merge(parent.parent)
+}
+
+// markFree internally moves a region with `offset` from busy to free map
+func (pa *PoolAllocator) markFree(memCls classType, offset uint64) (*region, error) {
+	clsPool := pa.pools[memCls]
+	if clsPool == nil {
+		return nil, ErrEmptyPoolOperation
+	}
+
+	if reg, exists := clsPool.busy[offset]; exists {
+		clsPool.free[offset] = reg
+		delete(clsPool.busy, offset)
+		return reg, nil
+	}
+	return nil, ErrNotAllocated
+}
+
+// markBusy internally moves a region with `offset` from free to busy map
+func (pa *PoolAllocator) markBusy(memCls classType, offset uint64) (*region, error) {
+	clsPool := pa.pools[memCls]
+	if clsPool == nil {
+		return nil, ErrEmptyPoolOperation
+	}
+
+	if reg, exists := clsPool.free[offset]; exists {
+		clsPool.busy[offset] = reg
+		delete(clsPool.free, offset)
+		return reg, nil
+	}
+	return nil, ErrNotAllocated
+}
+
+// Busy checks if there are any memory regions that are marked as busy
+func (pa *PoolAllocator) Busy() bool {
+	for i := classType(0); i < memoryClassNumber; i++ {
+		if len(pa.pools[i].busy) > 0 {
+			return false
```

Comment:
Shouldn't we return `true` here since we found at least one busy pool?

Reply:
good catch. but it also seems like this got refactored out, so I'll be removing this method, since I'm not using it anymore